kroxylicious-filters/kroxylicious-sasl-inspection/src/main/java/io/kroxylicious/filters/sasl/inspection/OauthBearerSaslObserver.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filters.sasl.inspection;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+
+import javax.security.sasl.SaslException;
+
+import org.apache.kafka.common.security.oauthbearer.internals.OAuthBearerClientInitialResponse;
+import org.jose4j.jwt.GeneralJwtException;
+import org.jose4j.jwt.consumer.InvalidJwtException;
+import org.jose4j.jwt.consumer.JwtConsumerBuilder;
+
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * A SASL observer capable of extracting an authorization id from a
+ * <a href="https://datatracker.ietf.org/doc/html/rfc7628">OAUTHBEARER</a> initial client response.
+ * <br/>
+ * The authorization id is obtained in the following manner.
+ * If the authorization id is present in the <a href="https://www.rfc-editor.org/rfc/rfc5801.html#page-8">gs2-header</a> it is used.
+ * If there is no authorization id in the header, the subject claim in the JWT payload is used.  If the subject
+ * claim is absent, the payload is encrypted (JWE), or opaque tokens are in use, the authentication will fail.
```

Comment:
This means we are getting involved in authn (though only in the negative case). I think that's probably OK for now, but it's a slightly complicated story.

Reply:
maybe we should add some docs at a higher level, like up on the `SaslInspectionFilter` explaining this other outcome where the SASL handshake succeeds but the Inspection Filter cannot extract an identity. We must close the connection because other Filters in the chain may expect an identity to be extracted.

Something that helps implementors of other SASL mechanisms understand the contract.